src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -508,6 +523,80 @@ mod tests {
         assert_eq!(right.schema().fields(), &fields[3..4]);
     }
 
+    /// Semi join panicked previously at `prune_col`. Add test to prevent regression.
+    #[tokio::test]
+    async fn test_semi_join_does_not_panic() {
+        let ty = DataType::Int32;
+        let ctx = OptimizerContext::mock().await;
+        let fields: Vec<Field> = (1..7)
+            .map(|i| Field::with_name(ty.clone(), format!("v{}", i)))
+            .collect();
+        let left = LogicalValues::new(
+            vec![],
+            Schema {
+                fields: fields[0..3].to_vec(),
+            },
+            ctx.clone(),
+        );
+        let right = LogicalValues::new(
+            vec![],
+            Schema {
+                fields: fields[3..6].to_vec(),
+            },
+            ctx,
+        );
+        let on: ExprImpl = ExprImpl::FunctionCall(Box::new(
+            FunctionCall::new(
+                Type::Equal,
+                vec![
+                    ExprImpl::InputRef(Box::new(InputRef::new(1, ty.clone()))),
+                    ExprImpl::InputRef(Box::new(InputRef::new(4, ty))),
+                ],
+            )
+            .unwrap(),
+        ));
+        for join_type in [
+            JoinType::LeftSemi,
+            JoinType::RightSemi,
+            JoinType::LeftAnti,
+            JoinType::RightAnti,
+        ] {
+            let join = LogicalJoin::new(
+                left.clone().into(),
+                right.clone().into(),
+                join_type,
+                Condition::with_expr(on.clone()),
+            );
+
+            let offset = if join.is_right_join() { 3 } else { 0 };
+
+            // Perform the prune
+            let mut required_cols = FixedBitSet::with_capacity(3);
+            // key 0 is never used in the join (always key 1)
+            required_cols.extend(vec![0]);
+            // should not panic here
+            let plan = join.prune_col(&required_cols);
+            // Check that the join has been wrapped in a projection
+            let as_plan = plan.as_logical_project().unwrap();
+            // Check the result
+            assert_eq!(as_plan.schema().fields().len(), 1);
+            assert_eq!(as_plan.schema().fields()[0], fields[offset + 0]);
```

Comment:
clippy failed on `offset + 0`. If you are sure it's more readable, you can add allow(clippy::?) attribute similar to https://stackoverflow.com/questions/55402812/how-to-disable-a-clippy-lint-for-a-single-line-block